backend/src/apiserver/server/pipeline_server.go

```diff
@@ -588,9 +589,9 @@ func NewPipelineServer(resourceManager *resource.ResourceManager, options *Pipel
 // Creates a pipeline version from. Not exported.
 // Applies common logic on v1beta1 and v2beta1 API.
 func (s *PipelineServer) createPipelineVersion(ctx context.Context, pv *model.PipelineVersion) (*model.PipelineVersion, error) {
-	// Fail if pipeline spec is missing
-	if (pv.PipelineSpec == "") && (pv.CodeSourceUrl == "") && (pv.PipelineSpecURI == "") {
-		return nil, util.NewInvalidInputError("Failed to create a pipeline version due to missing pipeline spec")
+	// Fail if pipeline URL is missing
+	if pv.PipelineSpecURI == "" {
```

Comment:
What happens when the pipeline is not provided through URL? For example, when it is uploaded directly.

Reply:
This is a great question.

Uploading is a special case where we parse the content of a request body into `PipelineSpec` field. We won't have a URL in that case and will directly call `ResourceManager`, which handles both cases.

Reply:
upload won't hit this code path, this `pipeline_server` only has `Create` methods, while `Upload` belongs to `pipeline_upload_server`: https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/pipeline_upload_server.go

Reply:
I see! Thank you for clarifying, I mixed up these two APIs.